File: modules/vision-camera-pose/android/src/main/java/com/margelo/nitro/newpic/visioncamerapose/HybridVisionCameraPoseFrameSink.kt
package com.margelo.nitro.newpic.visioncamerapose

import android.os.SystemClock
import android.util.Log
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import com.margelo.nitro.NitroModules
import com.margelo.nitro.camera.CameraOrientation
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.public.NativeFrame
import com.newpic.visioncamerapose.OwnedPoseInput
import com.newpic.visioncamerapose.PoseFrameInputAdapter
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.min

class HybridVisionCameraPoseFrameSink :
  HybridVisionCameraPoseFrameSinkSpec() {
  private val stateLock = Any()
  private val deliveryLock = Any()
  private val inferenceExecutor =
    Executors.newSingleThreadExecutor { runnable ->
      Thread(runnable, "NewpicPoseLandmarker")
    }
  private val isBusy = AtomicBoolean(false)

  private var options = ValidatedPoseOptions.defaults()
  private var poseLandmarker: PoseLandmarker? = null
  private var initializingGeneration: Long? = null
  private var acceptingFrames = false
  private var generation = 0L
  private var lastAcceptedTimestampMs = Long.MIN_VALUE
  private var lastMediaPipeTimestampMs = Long.MIN_VALUE
  private var pendingInput: PendingInput? = null

  private var resultCallback:
    ((NativeDetectedPoseFrame) -> Unit)? = null
  private var errorCallback:
    ((NativePoseDetectionError) -> Unit)? = null
  private var nextSequence = 1L
  private var inFlightSequence: Long? = null
  private var pendingResult: NativeDetectedPoseFrame? = null

  override fun configure(options: NativePoseDetectionOptions) {
    val validated = ValidatedPoseOptions.from(options)
    val shouldRebuild =
      synchronized(stateLock) {
        val changed = this.options != validated
        this.options = validated
        changed &&
          (
            poseLandmarker != null ||
              initializingGeneration != null
          )
      }

    if (shouldRebuild) {
      rebuildDetector()
    } else {
      ensureDetector()
    }
  }

  override fun startAcceptingFrames() {
    synchronized(stateLock) {
      acceptingFrames = true
    }
    ensureDetector()
  }

  override fun stopAcceptingFrames() {
    synchronized(stateLock) {
      acceptingFrames = false
      generation += 1
      lastAcceptedTimestampMs = Long.MIN_VALUE
    }
    clearPendingResults()
  }

  override fun releaseDetector() {
    val detector =
      synchronized(stateLock) {
        acceptingFrames = false
        generation += 1
        initializingGeneration = null
        lastAcceptedTimestampMs = Long.MIN_VALUE
        lastMediaPipeTimestampMs = Long.MIN_VALUE
        poseLandmarker.also {
          poseLandmarker = null
        }
      }
    clearPendingResults()
    inferenceExecutor.execute {
      detector?.close()
      finishPendingInput()
    }
  }

  override fun pushFrame(frame: HybridFrameSpec): Boolean {
    val nativeFrame = frame as? NativeFrame ?: return false
    val image = nativeFrame.image
    val frameTimestampMs =
      nextMediaPipeTimestampMs(
        image.imageInfo.timestamp / NANOSECONDS_PER_MILLISECOND,
      )

    val snapshot =
      synchronized(stateLock) {
        val detector = poseLandmarker ?: return false
        if (!acceptingFrames || isBusy.get()) {
          return false
        }
        val minimumIntervalMs =
          1000.0 / options.maxInferenceFps
        if (
          lastAcceptedTimestampMs != Long.MIN_VALUE &&
          frameTimestampMs - lastAcceptedTimestampMs <
          minimumIntervalMs
        ) {
          return false
        }
        if (!isBusy.compareAndSet(false, true)) {
          return false
        }
        lastAcceptedTimestampMs = frameTimestampMs
        DetectorSnapshot(
          detector = detector,
          generation = generation,
          options = options,
        )
      }

    val input =
      try {
        PoseFrameInputAdapter.copy(
          image = image,
          orientation = frame.orientation,
          isMirrored = frame.isMirrored,
          maxInputLongEdge =
            snapshot.options.maxInputLongEdge,
          timestampMs = frameTimestampMs,
        )
      } catch (error: Throwable) {
        isBusy.set(false)
        emitError("E_POSE_INPUT", error)
        return false
      }

    synchronized(stateLock) {
      pendingInput =
        PendingInput(
          input = input,
          generation = snapshot.generation,
        )
    }
    inferenceExecutor.execute {
      detect(snapshot, input)
    }
    return true
  }

  override fun setResultCallback(
    callback: ((NativeDetectedPoseFrame) -> Unit)?,
  ) {
    synchronized(deliveryLock) {
      resultCallback = callback
      if (callback == null) {
        inFlightSequence = null
        pendingResult = null
      }
    }
  }

  override fun setErrorCallback(
    callback: ((NativePoseDetectionError) -> Unit)?,
  ) {
    synchronized(deliveryLock) {
      errorCallback = callback
    }
  }

  override fun acknowledgeResult(sequence: Double) {
    val delivery =
      synchronized(deliveryLock) {
        if (inFlightSequence != sequence.toLong()) {
          return
        }
        inFlightSequence = null
        val next = pendingResult
        pendingResult = null
        if (next == null || resultCallback == null) {
          null
        } else {
          inFlightSequence = next.sequence.toLong()
          resultCallback to next
        }
      }
    delivery?.let { (callback, result) ->
      invokeResultCallback(callback, result)
    }
  }

  override fun dispose() {
    releaseDetector()
    synchronized(deliveryLock) {
      resultCallback = null
      errorCallback = null
    }
    inferenceExecutor.shutdown()
    super.dispose()
  }

  private fun ensureDetector() {
    val setup =
      synchronized(stateLock) {
        if (
          poseLandmarker != null ||
          initializingGeneration != null
        ) {
          return
        }
        generation += 1
        initializingGeneration = generation
        DetectorSetup(generation, options)
      }
    inferenceExecutor.execute {
      createDetector(setup)
    }
  }

  private fun rebuildDetector() {
    val setup =
      synchronized(stateLock) {
        generation += 1
        initializingGeneration = generation
        val old = poseLandmarker
        poseLandmarker = null
        DetectorSetup(generation, options, old)
      }
    inferenceExecutor.execute {
      createDetector(setup)
    }
  }

  private fun createDetector(setup: DetectorSetup) {
    try {
      setup.previousDetector?.close()
      finishPendingInput()
      val baseOptions =
        BaseOptions.builder()
          .setModelAssetPath(MODEL_ASSET_PATH)
          .build()
      val landmarkerOptions =
        PoseLandmarker.PoseLandmarkerOptions.builder()
          .setBaseOptions(baseOptions)
          .setRunningMode(RunningMode.LIVE_STREAM)
          .setNumPoses(setup.options.numPoses)
          .setMinPoseDetectionConfidence(
            setup.options.minPoseDetectionConfidence,
          )
          .setMinPosePresenceConfidence(
            setup.options.minPosePresenceConfidence,
          )
          .setMinTrackingConfidence(
            setup.options.minTrackingConfidence,
          )
          .setOutputSegmentationMasks(false)
          .setResultListener(::handleResult)
          .setErrorListener(::handleInferenceError)
          .build()
      val detector =
        PoseLandmarker.createFromOptions(
          requireNotNull(NitroModules.applicationContext) {
            "ReactApplicationContext is unavailable."
          }.applicationContext,
          landmarkerOptions,
        )

      val shouldKeep =
        synchronized(stateLock) {
          if (
            initializingGeneration == setup.generation
          ) {
            initializingGeneration = null
          }
          if (generation == setup.generation) {
            poseLandmarker = detector
            true
          } else {
            false
          }
        }
      if (!shouldKeep) {
        detector.close()
        ensureDetectorIfAccepting()
      }
    } catch (error: Throwable) {
      var shouldRetry = false
      synchronized(stateLock) {
        if (
          initializingGeneration == setup.generation
        ) {
          initializingGeneration = null
        }
        if (generation == setup.generation) {
          poseLandmarker = null
          isBusy.set(false)
        } else {
          shouldRetry =
            acceptingFrames &&
              poseLandmarker == null &&
              initializingGeneration == null
        }
      }
      emitError("E_POSE_INITIALIZATION", error)
      if (shouldRetry) {
        ensureDetector()
      }
    }
  }

  private fun detect(
    snapshot: DetectorSnapshot,
    input: OwnedPoseInput,
  ) {
    val isCurrent =
      synchronized(stateLock) {
        generation == snapshot.generation &&
          poseLandmarker === snapshot.detector
      }
    if (!isCurrent) {
      finishInput(input)
      return
    }

    try {
      val mpImage =
        BitmapImageBuilder(input.bitmap).build()
      snapshot.detector.detectAsync(
        mpImage,
        input.timestampMs,
      )
    } catch (error: Throwable) {
      finishInput(input)
      emitError("E_POSE_INFERENCE", error)
    }
  }

  private fun handleResult(
    result: PoseLandmarkerResult,
    inputImage: MPImage,
  ) {
    val pending =
      synchronized(stateLock) {
        pendingInput.also {
          pendingInput = null
        }
      }

    try {
      if (pending == null) {
        return
      }
      val input = pending.input
      val poses =
        result.landmarks().map { landmarks ->
          NativeDetectedPersonPose(
            landmarks =
              landmarks.map { landmark ->
                val visibility =
                  landmark.visibility().orElse(0f)
                val presence =
                  landmark.presence().orElse(0f)
                NativePosePoint(
                  x = landmark.x().toDouble(),
                  y = landmark.y().toDouble(),
                  z = landmark.z().toDouble(),
                  confidence =
                    min(visibility, presence).toDouble(),
                )
              }.toTypedArray(),
          )
        }.toTypedArray()

      val shouldDeliver =
        synchronized(stateLock) {
          acceptingFrames &&
            generation == pending.generation
        }
      if (shouldDeliver) {
        emitResult(
          timestampMs = result.timestampMs(),
          poses = poses,
          input = input,
        )
      }
    } finally {
      inputImage.close()
      pending?.input?.bitmap?.recycle()
      isBusy.set(false)
    }
  }

  private fun handleInferenceError(error: RuntimeException) {
    val input =
      synchronized(stateLock) {
        pendingInput.also {
          pendingInput = null
        }
      }
    input?.input?.bitmap?.recycle()
    isBusy.set(false)
    emitError("E_POSE_INFERENCE", error)
  }

  private fun finishInput(input: OwnedPoseInput) {
    synchronized(stateLock) {
      if (pendingInput?.input === input) {
        pendingInput = null
      }
    }
    input.bitmap.recycle()
    isBusy.set(false)
  }

  private fun finishPendingInput() {
    val pending =
      synchronized(stateLock) {
        pendingInput.also {
          pendingInput = null
        }
      }
    pending?.input?.bitmap?.recycle()
    isBusy.set(false)
  }

  private fun ensureDetectorIfAccepting() {
    val shouldEnsure =
      synchronized(stateLock) {
        acceptingFrames &&
          poseLandmarker == null &&
          initializingGeneration == null
      }
    if (shouldEnsure) {
      ensureDetector()
    }
  }

  private fun emitResult(
    timestampMs: Long,
    poses: Array<NativeDetectedPersonPose>,
    input: OwnedPoseInput,
  ) {
    val delivery =
      synchronized(deliveryLock) {
        val sequence = nextSequence++
        val frame =
          NativeDetectedPoseFrame(
            sequence = sequence.toDouble(),
            timestamp = timestampMs.toDouble(),
            poses = poses,
            inputWidth = input.inputWidth.toDouble(),
            inputHeight = input.inputHeight.toDouble(),
            sourceWidth = input.sourceWidth.toDouble(),
            sourceHeight = input.sourceHeight.toDouble(),
            rotationDegrees =
              input.rotationDegrees.toDouble(),
            isMirrored = input.isMirrored,
          )
        val callback = resultCallback
        if (callback == null) {
          return
        }
        if (inFlightSequence != null) {
          pendingResult = frame
          null
        } else {
          inFlightSequence = sequence
          callback to frame
        }
      }
    delivery?.let { (callback, frame) ->
      invokeResultCallback(callback, frame)
    }
  }

  private fun invokeResultCallback(
    callback: (NativeDetectedPoseFrame) -> Unit,
    frame: NativeDetectedPoseFrame,
  ) {
    try {
      callback(frame)
    } catch (error: Throwable) {
      Log.w(TAG, "Pose result callback is unavailable.", error)
      synchronized(deliveryLock) {
        if (inFlightSequence == frame.sequence.toLong()) {
          inFlightSequence = null
        }
      }
    }
  }

  private fun emitError(
    code: String,
    error: Throwable,
  ) {
    Log.e(TAG, code, error)
    val callback =
      synchronized(deliveryLock) {
        errorCallback
      } ?: return
    try {
      callback(
        NativePoseDetectionError(
          code = code,
          message =
            error.message ?: "Unknown MediaPipe Pose error.",
        ),
      )
    } catch (callbackError: Throwable) {
      Log.w(TAG, "Pose error callback is unavailable.", callbackError)
    }
  }

  private fun clearPendingResults() {
    synchronized(deliveryLock) {
      pendingResult = null
    }
  }

  private fun nextMediaPipeTimestampMs(
    sourceTimestampMs: Long,
  ): Long =
    synchronized(stateLock) {
      val source =
        if (sourceTimestampMs > 0) {
          sourceTimestampMs
        } else {
          SystemClock.uptimeMillis()
        }
      val timestamp =
        maxOf(source, lastMediaPipeTimestampMs + 1)
      lastMediaPipeTimestampMs = timestamp
      timestamp
    }

  private data class DetectorSnapshot(
    val detector: PoseLandmarker,
    val generation: Long,
    val options: ValidatedPoseOptions,
  )

  private data class PendingInput(
    val input: OwnedPoseInput,
    val generation: Long,
  )

  private data class DetectorSetup(
    val generation: Long,
    val options: ValidatedPoseOptions,
    val previousDetector: PoseLandmarker? = null,
  )

  private data class ValidatedPoseOptions(
    val numPoses: Int,
    val maxInferenceFps: Double,
    val maxInputLongEdge: Int,
    val minPoseDetectionConfidence: Float,
    val minPosePresenceConfidence: Float,
    val minTrackingConfidence: Float,
  ) {
    companion object {
      fun defaults() =
        ValidatedPoseOptions(
          numPoses = 4,
          maxInferenceFps = 10.0,
          maxInputLongEdge = 640,
          minPoseDetectionConfidence = 0.5f,
          minPosePresenceConfidence = 0.5f,
          minTrackingConfidence = 0.5f,
        )

      fun from(options: NativePoseDetectionOptions) =
        ValidatedPoseOptions(
          numPoses =
            options.numPoses.toInt().coerceIn(1, 4),
          maxInferenceFps =
            options.maxInferenceFps.coerceIn(1.0, 10.0),
          maxInputLongEdge =
            options.maxInputLongEdge.toInt()
              .coerceIn(256, 2048),
          minPoseDetectionConfidence =
            options.minPoseDetectionConfidence.toFloat()
              .coerceIn(0f, 1f),
          minPosePresenceConfidence =
            options.minPosePresenceConfidence.toFloat()
              .coerceIn(0f, 1f),
          minTrackingConfidence =
            options.minTrackingConfidence.toFloat()
              .coerceIn(0f, 1f),
        )
    }
  }

  companion object {
    private const val TAG = "VisionCameraPose"
    private const val MODEL_ASSET_PATH =
      "pose_landmarker_lite.task"
    private const val NANOSECONDS_PER_MILLISECOND = 1_000_000L
  }
}

File: modules/vision-camera-pose/ios/HybridVisionCameraPoseFrameSink.swift
import AVFoundation
import Foundation
import MediaPipeTasksVision
import NitroModules
import VisionCamera

private final class PoseLandmarkerDelegate:
  NSObject,
  PoseLandmarkerLiveStreamDelegate
{
  weak var owner: HybridVisionCameraPoseFrameSink?

  init(owner: HybridVisionCameraPoseFrameSink) {
    self.owner = owner
  }

  func poseLandmarker(
    _ poseLandmarker: PoseLandmarker,
    didFinishDetection result: PoseLandmarkerResult?,
    timestampInMilliseconds: Int,
    error: Error?
  ) {
    owner?.handleResult(
      result,
      timestampInMilliseconds: timestampInMilliseconds,
      error: error
    )
  }
}

final class HybridVisionCameraPoseFrameSink:
  HybridVisionCameraPoseFrameSinkSpec
{
  private let stateLock = NSLock()
  private let deliveryLock = NSLock()
  private let inferenceQueue = DispatchQueue(
    label: "com.newpic.vision-camera-pose.inference",
    qos: .userInitiated
  )
  private let inputAdapter = PoseFrameInputAdapter()

  private var options = ValidatedPoseOptions.defaults
  private var poseLandmarker: PoseLandmarker?
  private var landmarkerDelegate: PoseLandmarkerDelegate?
  private var initializingGeneration: Int64?
  private var acceptingFrames = false
  private var isBusy = false
  private var generation: Int64 = 0
  private var lastAcceptedTimestampMs: Int64 = .min
  private var lastMediaPipeTimestampMs: Int64 = .min
  private var pendingInput: PendingInput?

  private var resultCallback:
    ((NativeDetectedPoseFrame) -> Void)?
  private var errorCallback:
    ((NativePoseDetectionError) -> Void)?
  private var nextSequence: Int64 = 1
  private var inFlightSequence: Int64?
  private var pendingResult: NativeDetectedPoseFrame?

  func configure(
    options: NativePoseDetectionOptions
  ) throws {
    let validated = ValidatedPoseOptions(options)
    let shouldRebuild = stateLock.withLock {
      let changed = self.options != validated
      self.options = validated
      return changed &&
        (
          poseLandmarker != nil ||
            initializingGeneration != nil
        )
    }

    if shouldRebuild {
      rebuildDetector()
    } else {
      ensureDetector()
    }
  }

  func startAcceptingFrames() throws {
    stateLock.withLock {
      acceptingFrames = true
    }
    ensureDetector()
  }

  func stopAcceptingFrames() throws {
    stateLock.withLock {
      acceptingFrames = false
      generation &+= 1
      lastAcceptedTimestampMs = .min
    }
    clearPendingResults()
  }

  func releaseDetector() throws {
    var detector = stateLock.withLock {
      acceptingFrames = false
      generation &+= 1
      initializingGeneration = nil
      lastAcceptedTimestampMs = .min
      lastMediaPipeTimestampMs = .min
      let detector = poseLandmarker
      poseLandmarker = nil
      landmarkerDelegate = nil
      return detector
    }
    clearPendingResults()
    inferenceQueue.async { [weak self] in
      withExtendedLifetime(detector) {}
      detector = nil
      self?.finishPendingInput()
    }
  }

  func pushFrame(
    frame: any HybridFrameSpec
  ) throws -> Bool {
    guard
      let nativeFrame = frame as? any NativeFrame,
      let sampleBuffer = nativeFrame.sampleBuffer
    else {
      return false
    }

    let timestampMs = nextMediaPipeTimestampMs(
      sampleBuffer: sampleBuffer
    )
    guard let snapshot = stateLock.withLock({ () -> DetectorSnapshot? in
      guard
        acceptingFrames,
        !isBusy,
        let detector = poseLandmarker
      else {
        return nil
      }
      let minimumIntervalMs =
        1000.0 / options.maxInferenceFps
      if
        lastAcceptedTimestampMs != .min,
        Double(timestampMs - lastAcceptedTimestampMs) <
          minimumIntervalMs
      {
        return nil
      }
      isBusy = true
      lastAcceptedTimestampMs = timestampMs
      return DetectorSnapshot(
        detector: detector,
        generation: generation,
        options: options
      )
    }) else {
      return false
    }

    let input: OwnedPoseInput
    do {
      input = try inputAdapter.copy(
        sampleBuffer: sampleBuffer,
        orientation: frame.orientation,
        isMirrored: frame.isMirrored,
        maxInputLongEdge:
          snapshot.options.maxInputLongEdge,
        timestampMs: Int(timestampMs)
      )
    } catch {
      stateLock.withLock {
        isBusy = false
      }
      emitError(code: "E_POSE_INPUT", error: error)
      return false
    }

    stateLock.withLock {
      pendingInput = PendingInput(
        input: input,
        generation: snapshot.generation
      )
    }
    inferenceQueue.async { [weak self] in
      self?.detect(snapshot: snapshot, input: input)
    }
    return true
  }

  func setResultCallback(
    callback:
      ((NativeDetectedPoseFrame) -> Void)?
  ) throws {
    deliveryLock.withLock {
      resultCallback = callback
      if callback == nil {
        inFlightSequence = nil
        pendingResult = nil
      }
    }
  }

  func setErrorCallback(
    callback:
      ((NativePoseDetectionError) -> Void)?
  ) throws {
    deliveryLock.withLock {
      errorCallback = callback
    }
  }

  func acknowledgeResult(sequence: Double) throws {
    let delivery:
      (
        ((NativeDetectedPoseFrame) -> Void),
        NativeDetectedPoseFrame
      )? = deliveryLock.withLock {
        guard
          inFlightSequence == Int64(sequence)
        else {
          return nil
        }
        inFlightSequence = nil
        guard
          let next = pendingResult,
          let callback = resultCallback
        else {
          pendingResult = nil
          return nil
        }
        pendingResult = nil
        inFlightSequence = Int64(next.sequence)
        return (callback, next)
      }
    if let delivery {
      invokeResultCallback(
        delivery.0,
        frame: delivery.1
      )
    }
  }

  func dispose() {
    try? releaseDetector()
    deliveryLock.withLock {
      resultCallback = nil
      errorCallback = nil
    }
  }

  deinit {
    dispose()
  }

  fileprivate func handleResult(
    _ result: PoseLandmarkerResult?,
    timestampInMilliseconds: Int,
    error: Error?
  ) {
    let pending = stateLock.withLock {
      defer {
        pendingInput = nil
        isBusy = false
      }
      return pendingInput
    }

    if let error {
      emitError(code: "E_POSE_INFERENCE", error: error)
      return
    }
    guard
      let result,
      let pending
    else {
      emitError(
        code: "E_POSE_INFERENCE",
        error: PoseDetectionError.missingResult
      )
      return
    }

    let shouldDeliver = stateLock.withLock {
      acceptingFrames &&
        generation == pending.generation
    }
    guard shouldDeliver else {
      return
    }

    let poses = result.landmarks.map { landmarks in
      NativeDetectedPersonPose(
        landmarks: landmarks.map { landmark in
          let visibility =
            landmark.visibility?.doubleValue ?? 0
          let presence =
            landmark.presence?.doubleValue ?? 0
          return NativePosePoint(
            x: Double(landmark.x),
            y: Double(landmark.y),
            z: Double(landmark.z),
            confidence: min(visibility, presence)
          )
        }
      )
    }
    emitResult(
      timestampMs: timestampInMilliseconds,
      poses: poses,
      input: pending.input
    )
  }

  private func ensureDetector() {
    let setup: DetectorSetup? = stateLock.withLock {
      guard
        poseLandmarker == nil,
        initializingGeneration == nil
      else {
        return nil
      }
      generation &+= 1
      initializingGeneration = generation
      return DetectorSetup(
        generation: generation,
        options: options
      )
    }
    guard let setup else {
      return
    }
    inferenceQueue.async { [weak self] in
      self?.createDetector(setup: setup)
    }
  }

  private func rebuildDetector() {
    var previousDetector: PoseLandmarker?
    let setup = stateLock.withLock {
      generation &+= 1
      initializingGeneration = generation
      isBusy = false
      previousDetector = poseLandmarker
      poseLandmarker = nil
      landmarkerDelegate = nil
      return DetectorSetup(
        generation: generation,
        options: options
      )
    }
    inferenceQueue.async { [weak self] in
      withExtendedLifetime(previousDetector) {}
      previousDetector = nil
      self?.finishPendingInput()
      self?.createDetector(setup: setup)
    }
  }

  private func createDetector(setup: DetectorSetup) {
    do {
      let modelPath = try Self.modelPath()
      let delegate = PoseLandmarkerDelegate(owner: self)
      let landmarkerOptions = PoseLandmarkerOptions()
      landmarkerOptions.baseOptions.modelAssetPath = modelPath
      landmarkerOptions.runningMode = .liveStream
      landmarkerOptions.numPoses = setup.options.numPoses
      landmarkerOptions.minPoseDetectionConfidence =
        setup.options.minPoseDetectionConfidence
      landmarkerOptions.minPosePresenceConfidence =
        setup.options.minPosePresenceConfidence
      landmarkerOptions.minTrackingConfidence =
        setup.options.minTrackingConfidence
      landmarkerOptions.shouldOutputSegmentationMasks = false
      landmarkerOptions.poseLandmarkerLiveStreamDelegate =
        delegate
      let detector = try PoseLandmarker(
        options: landmarkerOptions
      )

      let shouldKeep = stateLock.withLock {
        if initializingGeneration == setup.generation {
          initializingGeneration = nil
        }
        guard generation == setup.generation else {
          return false
        }
        poseLandmarker = detector
        landmarkerDelegate = delegate
        return true
      }
      if !shouldKeep {
        ensureDetectorIfAccepting()
      }
    } catch {
      var shouldRetry = false
      stateLock.withLock {
        if initializingGeneration == setup.generation {
          initializingGeneration = nil
        }
        if generation == setup.generation {
          poseLandmarker = nil
          landmarkerDelegate = nil
          isBusy = false
        } else {
          shouldRetry =
            acceptingFrames &&
              poseLandmarker == nil &&
              initializingGeneration == nil
        }
      }
      emitError(
        code: "E_POSE_INITIALIZATION",
        error: error
      )
      if shouldRetry {
        ensureDetector()
      }
    }
  }

  private func detect(
    snapshot: DetectorSnapshot,
    input: OwnedPoseInput
  ) {
    let isCurrent = stateLock.withLock {
      generation == snapshot.generation &&
        poseLandmarker === snapshot.detector
    }
    guard isCurrent else {
      finishPendingInput()
      return
    }

    do {
      let image = try MPImage(
        pixelBuffer: input.pixelBuffer
      )
      try snapshot.detector.detectAsync(
        image: image,
        timestampInMilliseconds: input.timestampMs
      )
    } catch {
      finishPendingInput()
      emitError(code: "E_POSE_INFERENCE", error: error)
    }
  }

  private func finishPendingInput() {
    stateLock.withLock {
      pendingInput = nil
      isBusy = false
    }
  }

  private func ensureDetectorIfAccepting() {
    let shouldEnsure = stateLock.withLock {
      acceptingFrames &&
        poseLandmarker == nil &&
        initializingGeneration == nil
    }
    if shouldEnsure {
      ensureDetector()
    }
  }

  private func emitResult(
    timestampMs: Int,
    poses: [NativeDetectedPersonPose],
    input: OwnedPoseInput
  ) {
    let delivery:
      (
        ((NativeDetectedPoseFrame) -> Void),
        NativeDetectedPoseFrame
      )? = deliveryLock.withLock {
        let sequence = nextSequence
        nextSequence &+= 1
        let frame = NativeDetectedPoseFrame(
          sequence: Double(sequence),
          timestamp: Double(timestampMs),
          poses: poses,
          inputWidth: Double(input.inputWidth),
          inputHeight: Double(input.inputHeight),
          sourceWidth: Double(input.sourceWidth),
          sourceHeight: Double(input.sourceHeight),
          rotationDegrees: Double(input.rotationDegrees),
          isMirrored: input.isMirrored
        )
        guard let callback = resultCallback else {
          return nil
        }
        if inFlightSequence != nil {
          pendingResult = frame
          return nil
        }
        inFlightSequence = sequence
        return (callback, frame)
      }
    if let delivery {
      invokeResultCallback(
        delivery.0,
        frame: delivery.1
      )
    }
  }

  private func invokeResultCallback(
    _ callback:
      @escaping (NativeDetectedPoseFrame) -> Void,
    frame: NativeDetectedPoseFrame
  ) {
    callback(frame)
  }

  private func emitError(
    code: String,
    error: Error
  ) {
    let callback = deliveryLock.withLock {
      errorCallback
    }
    callback?(
      NativePoseDetectionError(
        code: code,
        message: error.localizedDescription
      )
    )
  }

  private func clearPendingResults() {
    deliveryLock.withLock {
      pendingResult = nil
    }
  }

  private func nextMediaPipeTimestampMs(
    sampleBuffer: CMSampleBuffer
  ) -> Int64 {
    let presentationTime =
      CMSampleBufferGetPresentationTimeStamp(sampleBuffer)
    let seconds = CMTimeGetSeconds(presentationTime)
    let sourceTimestamp =
      seconds.isFinite && seconds >= 0
      ? Int64((seconds * 1000).rounded(.down))
      : Int64(
        DispatchTime.now().uptimeNanoseconds /
          1_000_000
      )

    return stateLock.withLock {
      let timestamp = max(
        sourceTimestamp,
        lastMediaPipeTimestampMs &+ 1
      )
      lastMediaPipeTimestampMs = timestamp
      return timestamp
    }
  }

  private static func modelPath() throws -> String {
    let hostBundle = Bundle(
      for: PoseResourceLocator.self
    )
    let bundleURL =
      hostBundle.url(
        forResource: "VisionCameraPoseResources",
        withExtension: "bundle"
      ) ??
      Bundle.main.url(
        forResource: "VisionCameraPoseResources",
        withExtension: "bundle"
      )
    guard
      let bundleURL,
      let resourceBundle = Bundle(url: bundleURL),
      let path = resourceBundle.path(
        forResource: "pose_landmarker_lite",
        ofType: "task"
      )
    else {
      throw PoseDetectionError.modelNotFound
    }
    return path
  }
}

private final class PoseResourceLocator: NSObject {}

private struct PendingInput {
  let input: OwnedPoseInput
  let generation: Int64
}

private struct DetectorSnapshot {
  let detector: PoseLandmarker
  let generation: Int64
  let options: ValidatedPoseOptions
}

private struct DetectorSetup {
  let generation: Int64
  let options: ValidatedPoseOptions
}

private struct ValidatedPoseOptions: Equatable {
  let numPoses: Int
  let maxInferenceFps: Double
  let maxInputLongEdge: Int
  let minPoseDetectionConfidence: Float
  let minPosePresenceConfidence: Float
  let minTrackingConfidence: Float

  static let defaults = ValidatedPoseOptions(
    numPoses: 4,
    maxInferenceFps: 10,
    maxInputLongEdge: 640,
    minPoseDetectionConfidence: 0.5,
    minPosePresenceConfidence: 0.5,
    minTrackingConfidence: 0.5
  )

  init(_ options: NativePoseDetectionOptions) {
    numPoses = min(max(Int(options.numPoses), 1), 4)
    maxInferenceFps = min(
      max(options.maxInferenceFps, 1),
      10
    )
    maxInputLongEdge = min(
      max(Int(options.maxInputLongEdge), 256),
      2048
    )
    minPoseDetectionConfidence = Float(
      min(max(options.minPoseDetectionConfidence, 0), 1)
    )
    minPosePresenceConfidence = Float(
      min(max(options.minPosePresenceConfidence, 0), 1)
    )
    minTrackingConfidence = Float(
      min(max(options.minTrackingConfidence, 0), 1)
    )
  }

  private init(
    numPoses: Int,
    maxInferenceFps: Double,
    maxInputLongEdge: Int,
    minPoseDetectionConfidence: Float,
    minPosePresenceConfidence: Float,
    minTrackingConfidence: Float
  ) {
    self.numPoses = numPoses
    self.maxInferenceFps = maxInferenceFps
    self.maxInputLongEdge = maxInputLongEdge
    self.minPoseDetectionConfidence =
      minPoseDetectionConfidence
    self.minPosePresenceConfidence =
      minPosePresenceConfidence
    self.minTrackingConfidence = minTrackingConfidence
  }
}

private enum PoseDetectionError: LocalizedError {
  case modelNotFound
  case missingResult

  var errorDescription: String? {
    switch self {
    case .modelNotFound:
      return "pose_landmarker_lite.task is missing."
    case .missingResult:
      return "MediaPipe Pose returned no result."
    }
  }
}

private extension NSLock {
  func withLock<T>(_ body: () throws -> T) rethrows -> T {
    lock()
    defer { unlock() }
    return try body()
  }
}
